lib/active_merchant/billing/gateways/ixopay.rb

```diff
@@ -67,55 +54,193 @@ def scrub(transcript)
 
       private
 
-      def add_customer_data(post, options)
+      def headers(xml)
+        timestamp = Time.now.httpdate
+        signature = generate_signature('POST', xml, timestamp)
+
+        {
+          'Authorization' => "Gateway #{options[:api_key]}:#{signature}",
+          'Date' => timestamp,
+          'Content-Type' => 'text/xml; charset=utf-8'
+        }
+      end
+
+      def generate_signature(http_method, xml, timestamp)
+        content_type = 'text/xml; charset=utf-8'
+
+        message = http_method + "\n" + Digest::MD5.hexdigest(xml) + "\n" + content_type + "\n" \
+          + timestamp + "\n\n" + '/transaction'
+
+        digest = OpenSSL::Digest.new('sha512')
+        hmac = OpenSSL::HMAC.digest(digest, @secret, message)
+
+        Base64.encode64(hmac).split.join
+      end
+
+      def parse(action, xml)
+        parse_element({:action => action}, REXML::Document.new(xml))
+      end
+
+      # This generic method appears in a number of gateways that parse XML.
+      # In the future, we should investigate finding a library method to
+      # drop in, or factoring it out to a module or base class.
+      def parse_element(raw, node)
+        node_name = node.name.underscore
+
+        node.attributes.each do |k, v|
+          raw["#{node_name}_#{k.underscore}".to_sym] = v
+        end
+
+        if node.has_elements?
+          raw[node_name.to_sym] = true unless node.name.blank?
+          node.elements.each { |e| parse_element(raw, e) }
+        elsif node.children.count > 1
+          raw[node_name.to_sym] = node.children.join(' ').strip
+        else
+          raw[node_name.to_sym] = node.text unless node.text.nil?
+        end
+
+        raw
+      end
+
+      def build_purchase_request(money, payment_method, options)
+        xml = Builder::XmlMarkup.new(indent: 2)
+
+        xml.instruct! :xml, encoding: 'utf-8'
```

Comment:
It appears that `xml.instruct!` includes UTF-8 encoding by default and doesn't need to be specified: https://apidock.com/rails/v1.0.0/Builder/XmlMarkup/instruct%21

Reply:
Removed.